Pauli exponential support

radumarg opened the pull request:
The primary motivation for adding native Pauli‐exponential (`pauliexp`) support is performance - when we send `pauliexp` directly to IonQ, our compiler can optimize across a larger context (instead of first Trotterizing), which speeds up development and execution times. No existing workflows are strictly blocked, but allowing `pauliexp` to go through "as-is" reduces compilation overhead and yields shorter circuits.

`pauliexp` isn’t yet documented publicly, we already support it. The qiskit‐ionq package already includes helpers for Pauli exponentials ([Qiskit‐IonQ helpers](https://qiskit-community.github.io/qiskit-ionq/_modules/qiskit_ionq/helpers.html), so the hardware/runtime will accept it natively.

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -319,30 +332,286 @@ def _apply_operation(self, operation, circuit_index=0):
+                    # (3) Here I assume the hamiltonian depends on some  unspecified parametrs and one more argument wich is time. I do not know
+                    #     how to enforce this choice in API as the user may come up with a different parametrization, live having time as the first 
+                    #     argument for example.
+                    # (4) It is not clear how the number of steps for trotterization should be provided as input.
```
`ParametrizedEvolution` is not decomposed by Trotterization. In a simulation setting, we use an ODE solver to time-evolve the state vector based on the time steps. There is no way to specify the number of Trotter steps when using a `ParametrizedEvolution`, so the way to apply it to a QPU is not immediately clear to me. I will get input from the team to give you a thorough response.

mudit2812 reviewed (changes requested):
Thanks for working on this @radumarg . I'm going to do another review soon where I do an even deeper dive into your implementation. Right now, the review mostly covers stylistic and best practices suggestions.

I've also tried to address your concerns about the `ParametrizedEvolution` implementation, for which I'm also going to get more feedback from the team. Please don't hesitate to reach out for questions 😄

mudit2812 commented:
Hi @radumarg , could you please share the API docs that show `pauliexp` as a valid gate in the IonQ gate set? We used [this reference](https://docs.ionq.com/api-reference/v0.3/writing-quantum-programs) to get information about the gate set, but it does not mention `pauliexp` anywhere.

Additionally, I wanted to note that PennyLane has its own decomposition pipeline to decompose circuits into a gate set supported by the device as part of preprocessing. Before this PR, `Evolution` gates should have been decomposing via the Trotterization approximation automatically since it wasn't part of the device's gate set. Did executing a circuit containing the `Evolution` gate work before the changes in this PR?

radumarg commented:
@mudit2812 the Evolution gate does not seem to work on IonQ simulator: 
![image](https://github.com/user-attachments/assets/db7aa581-f796-4048-9aa4-00ec917f82eb)

radumarg commented:
@mudit2812 Regarding the documentation for pauliexp gate, Spencer Churchill @splch should be able to respond

Alex-Preciado commented:
Hi @radumarg, thanks for opening this PR 🚀 . Could you please add a brief description outlining the motivation behind the change and what it’s intended to do? This helps reviewers understand the context and ensures the project history remains clear for others reading it later on. Let me know if you have any questions.

radumarg commented:
@mudit2812 I have checked and using a circuit with QNode instead of QuantumTape works when jobs are sent to IonQ simulator: 
![image](https://github.com/user-attachments/assets/f3917473-db88-4a94-9fe2-9ce15a4733bc)

radumarg commented:
@mudit2812, @Alex-Preciado I am waiting for some feedback from the IonQ team so that I can provide a better description and context for this PR. Will get back when I have something more concrete.

radumarg commented on pennylane_ionq/exceptions.py:
```diff
@@ -0,0 +1,52 @@
+    """Raised when after submitting multiple circuits circuit index is not set
+    before the user want to access implementation methods of IonQDevice
+    like probability(), estimate_probability(), sample() or the prob property.
+    """
```
fixed

radumarg commented on pennylane_ionq/exceptions.py:
```diff
@@ -0,0 +1,52 @@
+
+    def __init__(self):
+        self.message = (
+            "The current instance of ParametrizedEvolution gate is not supported."
```
implementation of ParametrizedEvolution was removed

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -319,30 +332,286 @@ def _apply_operation(self, operation, circuit_index=0):
+            ops = pauli_rep.ops
+        elif isinstance(operation_generator, SProd):
+            if isinstance(operation_generator.base, (PauliX, PauliY, PauliZ, Identity)):
+                # TODO: single Pauli exp not working at IonQ?
```
There was a bug in IonQ API which was meanwhile fixed. I removed the comment.

radumarg commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -0,0 +1,475 @@
+
+        assert np.allclose(
+            result_ionq, results_qiskit_statevector_sim, atol=1e-2
+        ), "The IonQ and simulator results do not agree."
```
For various reasons your suggestion does not work well for all tests (there are more differences than input parameters). However I implemented this pattern here: test_evolution_object_created_from_hamiltonian

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -414,7 +685,9 @@ def probability(self, wires=None, shot_range=None, bin_size=None):
             return self.marginal_prob(self.prob, wires)
 
-        return self.estimate_probability(wires=wires, shot_range=shot_range, bin_size=bin_size)
+        return self.estimate_probability(
+            wires=wires, shot_range=shot_range, bin_size=bin_size
+        )
```
will fix reformatting at the end

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -395,7 +664,9 @@ def prob(self):
         # Here, we rearrange the states to match the big-endian ordering
         # expected by PennyLane.
-        basis_states = (int(bin(int(k))[2:].rjust(self.num_wires, "0")[::-1], 2) for k in histogram)
+        basis_states = (
+            int(bin(int(k))[2:].rjust(self.num_wires, "0")[::-1], 2) for k in histogram
+        )
```
will fix reformatting at the end

radumarg commented on pennylane_ionq/device.py:
Fixed, thanks!

radumarg commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -0,0 +1,451 @@
+            result_ionq, result_simulator, atol=1e-2
+        ), "The IonQ and simulator results do not agree."
+
+    def test_evolution_object_created_from_prod(self, requires_api):
```
Implemented as suggested.

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -330,87 +319,53 @@ def _apply_operation(self, operation, circuit_index=0):
+            gate["terms"] = terms
+            gate["coefficients"] = [-1 * float(v) for v in coefficients]
+            gate["time"] = operation.param
+            self.input["circuits"][circuit_index]["circuit"].append(gate)
```
After discussing with Pennylane team I understood that tottering is handled internally by their implementation of pauliexp gates, this is why I removed this feature. I will double-check and return here with a confirmation.

radumarg replied:
Update (July 2nd 2025) we are still discussing this issue .. will get back later

mudit2812 replied:
Hey @radumarg , I discussed with the team, and here are our thoughts:
* In an ideal world, option 1 would be preferred, as it terminates execution before a user runs a potentially incorrect and _paid_ job. But, since there is no public facing documentation for `pauliexp` or the `IONQ_RESYNTHESIS` environment variable, option 1 does not make sense.
* Currently, option 2 seems like the best course of action. We raise a warning if a user uses `qml.evolve` that their specified `num_steps` will be ignored, and that the backend will automatically choose a compilation scheme. Note that this warning should be raised even if `num_steps` is not specified, since in base PennyLane, `num_steps=None` is equivalent to an exact time evolution, not an approximation. Along with the warning, this behaviour should be documented in the docstring of the device being updated.

Let me know if you have any questions. @isaacdevlugt feel free to chime in if I missed anything.

radumarg commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -0,0 +1,451 @@
+            result_ionq, result_simulator, atol=1e-2
+        ), "The IonQ and simulator results do not agree."
+
+    def test_evolution_object_created_from_exp_operator(self, requires_api):
+        """Test that the implementation of Evolution gate
+        derived from a Hamiltonian constructed via an Exp matrix."""
+
+        dev = qml.device("ionq.simulator", wires=2, gateset="qis")
+
+        H = 2.5 * qml.PauliX(0) + 0.5 * qml.PauliY(1)
+        t = 3
+        U = Exp(t * H)
+
+        time = 2
+        tape = qml.tape.QuantumScript([qml.evolve(U, time)], [qml.probs(wires=[0, 1])])
+
+        result_ionq = dev.batch_execute([tape])
+
+        simulator = qml.device("default.qubit", wires=2)
+        result_simulator = qml.execute([tape], simulator)
+
+        assert np.allclose(
+            result_ionq, result_simulator, atol=1e-2
+        ), "The IonQ and simulator results do not agree."
```
These are indeed duplicate tests. Removed one of them.

radumarg commented:
@mudit2812 I have implemented most of your suggestions, increased code coverage, which should now be complete for the code modified by this ticket. I have made some comments and added some sample Qiskit code above on the issue of Pennylane apparently not returning correct simulation results. Can you please have a look with your team and let me know what you find out?

mudit2812 commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -0,0 +1,451 @@
+
+        result_ionq = dev.batch_execute([tape])
+
+        # Pennylane simulator returns incorrect
+        # results for this test, probably because Pauli
+        # strings in which the operator is decomposed
+        # do not commute.
```
I also tried it with `lightning.qubit`, which does not natively support the `Evolution` gate without trotterization. With 500 trotter steps, the result converged to the same values I got with `default.qubit`.

mudit2812 replied:
Hi @radumarg . I decided to try another circuit, where I manually exponentiated the matrix using `scipy.linalg.expm`, and used the matrix as input to `qml.QubitUnitary`. This would have forced the PennyLane simulators to use the matrix for evolving the state rather than decomposing using trotterization, and I still get the same results as what I found in my previous message:
```python
import numpy as np
import pennylane as qml
from scipy.linalg import expm

matrix = np.array([
[1.0, 2.0, 0.0, 1.0],
[0.0, -1.0, 3.0, 0.0],
[0.0, 1.0, -1.0, 1.0],
[-1.0, -2.0, 0.0, 7.0],
], dtype=complex)

hermitian_matrix = (matrix + matrix.conj().T) / 2
mat_with_coeffs = hermitian_matrix * -1j * 3
exp_mat = expm(mat_with_coeffs)

@qml.qnode(qml.device("default.qubit"))
def circuit4():
    qml.QubitUnitary(exp_mat, wires=[0, 1])
    return qml.probs()

@qml.qnode(qml.device("lightning.qubit"))
def circuit5():
    qml.QubitUnitary(exp_mat, wires=[0, 1])
    return qml.probs()
```
```pycon
>>> print(circuit4())
[0.19460719 0.42143554 0.37709081 0.00686647]
>>> print(circuit5())
[0.19460719 0.42143554 0.37709081 0.00686647]
```

Since I'm using the matrix manually to evolve the state, at this point, I'm wondering if Qiskit is correct or not. I will continue investigating to ensure that we reach resolution, but I wanted to let you know as well.

radumarg replied:
@mudit2812 now that I think about it, this might be a qubit ordering issue. I will run some tests. Please do not spend more time on this just yet.

mudit2812 replied:
@radumarg the team was able to identify the source of this issue. Qiskit's default for synthesizing a `PauliEvolutionGate` is to use Trotterization with a single Trotter step. Meanwhile, PL simulators use the exact evolution rather than an approximation. If I run the same circuit in PL with `num_steps=1`, I get the same output as Qiskit:
```python
@qml.qnode(qml.device("default.qubit"))
def circuit1():
    qml.evolve(H, coeff=3, num_steps=1)
    return qml.probs()
```
```pycon
>>> print(circuit1())
[0.19460719 0.42143554 0.37709081 0.00686647]
```

The source of the issue is just the fact that we use different defaults than Qiskit does. I've reached out to the product team to get resolution on what makes the most sense as a default (use exact evolution, or trotterize with one step), and will reach out once I have feedback. Tomorrow is likely the earliest I'll be able to get back to you.

However, to unblock testing, you should be able to specify `num_steps=1` in the simulator circuit for validating your results.

radumarg replied:
@mudit2812 Thanks for your help! I will specify num_steps=1 in the simulator as you suggested.

mudit2812 reviewed (commented):
Totally forgot to request you to add an entry to the changelog 😅 . Thanks

radumarg commented:
@mudit2812 I added a warning specifying that the number of steps argument to the Evolution  gate will be ignored by the API and that the backend will automatically choose a compilation scheme. I added a test to cover this change.

There was another issue with the Pennylane simulation results not being the same to IonQ trotterized results due to  Pennylane calculating exact results instead of trotterized ones. I have tried your suggestion to use lightning.qubit but it does not seem to work for the two tests having this issue: test_evolution_object_created_from_sparse_hamiltonian_2 and test_evolution_object_created_from_hermitian_matrix_1. Here is the error I get: 

_FAILED tests/test_pauliexp_ionq_gate.py::TestIonQPauliexp::test_evolution_object_created_from_sparse_hamiltonian_2 - pennylane.DeviceError: Operator Evolution(-3j SparseHamiltonian) not supported with lightning.qubit and does not provide a decomposition._

In case there are other changes required please let me know. Otherwise I believe the work on this PR is complete.

mudit2812 commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -0,0 +1,426 @@
+        """Test that the implementation of Evolution gate derived
+        from a Hamiltonian constructed via an SProd term works.
+        """
+
+        dev = qml.device("ionq.simulator", wires=2, gateset="qis")
+
+        H1 = qml.Hamiltonian([1.0], [qml.PauliX(0)])
+        H2 = qml.Hamiltonian([-0.5], [qml.PauliZ(1)])
+        sum_H = 2 * SProd(2.5, H1) + SProd(3.1, H2)
+
+        time = 3
+        tape = qml.tape.QuantumScript([qml.evolve(sum_H, time)], [qml.probs(wires=[0, 1])])
+
+        result_ionq = dev.batch_execute([tape])
+
+        simulator = qml.device("default.qubit", wires=2)
+        result_simulator = qml.execute([tape], simulator)
+
+        assert np.allclose(
+            result_ionq, result_simulator, atol=1e-2
+        ), "The IonQ and simulator results do not agree."
+
+    def test_evolution_object_created_from_sprod_2(self, requires_api):
+        """Test that the implementation of Evolution gate derived
+        from a Hamiltonian constructed via an SProd term works.
+        """
+
+        dev = qml.device("ionq.simulator", wires=1, gateset="qis")
+
+        H = SProd(2.5, qml.PauliX(0))
+        tape = qml.tape.QuantumScript([qml.evolve(H)], [qml.probs(wires=[0])])
+
+        result_ionq = dev.batch_execute([tape])
+
+        simulator = qml.device("default.qubit", wires=2)
+        result_simulator = qml.execute([tape], simulator)
+
+        assert np.allclose(
+            result_ionq, result_simulator, atol=1e-2
+        ), "The IonQ and simulator results do not agree."
```
This test can be parametrized like shown below. The same strategy can be used to combine `test_evolution_object_created_from_hermitian_matrix_1` and `test_evolution_object_created_from_hermitian_matrix_2` as well.
```suggestion
    @pytest.mark.parametrize(
        "op", 
        [
            SProd(2.5, qml.PauliX(0)),
            2 * SProd(2.5, qml.Hamiltonian([1.0], [qml.PauliX(0)]))
            + SProd(3.1, qml.Hamiltonian([-0.5], [qml.PauliZ(1)])),
        ],
    )
    def test_evolution_object_created_from_sprod(self, op, requires_api):
        """Test that the implementation of Evolution gate derived
        from a Hamiltonian constructed via an SProd term works.
        """

        dev = qml.device("ionq.simulator", wires=2, gateset="qis")

        time = 3
        tape = qml.tape.QuantumScript([qml.evolve(op, time)], [qml.probs(wires=[0, 1])])

        result_ionq = dev.batch_execute([tape])

        simulator = qml.device("default.qubit", wires=2)
        result_simulator = qml.execute([tape], simulator)

        assert np.allclose(
            result_ionq, result_simulator, atol=1e-2
        ), "The IonQ and simulator results do not agree."
```

radumarg replied:
working on it, partially implemented, but have some issues with the simulation on lightning devices, see my comment below

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -318,11 +318,40 @@ def _apply_operation(self, operation, circuit_index=0):
+        The number of steps argument for Evolution gate will be ignored even if provided because
+        IonQ implements hardware-efficient approximate compilation schemes for pauliexp gates.
+        """
+        warnings.warn(
+            "The number of steps argument for Evolution gate will be ignored even if provided because "
+            "IonQ implements hardware-efficient approximate compilation schemes for pauliexp gates "
+            "and the Evolution gate is being mapped to pauliexp gates by the API before sending to IonQ.",
+            UserWarning,
+        )
```
cc @isaacdevlugt

mudit2812 reviewed (commented):
Thanks for adding the warning @radumarg . Regarding the test failure you mentioned, the failure should be resolved once you add `num_steps=1` to the evolution op when executing on `lightning.qubit`. I'm ready to approve once that change is made instead of using a manually defined probability distribution.

I've also left some final clean-up comments.

mudit2812 commented:
@radumarg in terms of test coverage, the tests look good to me. Could you confirm that the tests give you 100% coverage of your changes? You can do so by installing the `pytest-cov` package and running the following command:
```
pytest <test file/directory> --cov-report term-missing --cov=<folder for which you want coverage information>
```

This will give you a coverage report in the terminal, and will also include details about any lines that are missed.

mudit2812 commented:
Hey @radumarg , sorry about that. There are two mistakes with the suggestion I gave you:
* The decomposition pathway for `Evolution` is not compatible with `qml.Hermitian`. You can create a Pauli sentence that PennyLane can use using `qml.pauli_decompose(hermitian_matrix)`
* We're in the process of deprecating the `num_steps` argument of `qml.evolve`, so for the simulator circuit, you should use `qml.TrotterProduct(H, time, n=1)` instead of `qml.evolve(H, time, num_steps=1)`.

Sorry for the confusion!

radumarg commented:
@mudit2812 I implemented your suggestions. Now the code does bot throw anymore but the simulation results do not agree with Qiskit. As you already know, the simulation results from running with IonQ simulator do agree with Qiskit. Here is my code. Can you please have a look:

        dev = qml.device("ionq.simulator", wires=2, gateset="qis")

        H_matrix = np.array(
            [[1, 1 + 1j, 0, -1j], [1 - 1j, 3, 2, 0], [0, 2, 0, 1j], [1j, 0, -1j, 1]]
        )
        hermitian_op = qml.Hermitian(H_matrix, wires=[0, 1])
        H = qml.Hamiltonian([2.0], [hermitian_op])

        time = 7
        
        # run on IonQ simulator
        #tape = qml.tape.QuantumScript([qml.evolve(H, time)], [qml.probs(wires=[0, 1])])
        #result_ionq = dev.batch_execute([tape])

        pauli_decomp = qml.pauli_decompose(H.matrix())
        tape = qml.tape.QuantumScript([qml.TrotterProduct(pauli_decomp, time, n=1)], [qml.probs(wires=[0, 1])])
        simulator = qml.device("lightning.qubit", wires=2)
        result_simulator = qml.execute([tape], simulator)

        results_qiskit_statevector_sim = [
            0.10784311,
            0.45583129,
            0.09056136,
            0.34576424,
        ]

        assert np.allclose(
            result_simulator, results_qiskit_statevector_sim, atol=1e-2
        )

mudit2812 commented:
Hey @radumarg, thanks for catching this, and sorry for the trouble! There are two things that are the source of failure, boiling down to differences in convention:
* Firstly, `qml.evolve` implements $\exp(-iHt)$, while `qml.TrotterProduct` implements $\exp(iHt)$.
* Secondly, `qml.TrotterProduct`'s decomposition reverses the order of the terms. This is not inherently incorrect, but it is a different convention to what `qml.evolve` and Qiskit use.

The below code example includes changes that accommodates for both of the aforementioned differences in convention:
```python
    dev = qml.device("ionq.simulator", wires=2, gateset="qis")

    H_matrix = np.array(
        [[1, 1 + 1j, 0, -1j], [1 - 1j, 3, 2, 0], [0, 2, 0, 1j], [1j, 0, -1j, 1]]
    )
    hermitian_op = qml.Hermitian(H_matrix, wires=[0, 1])
    H = qml.Hamiltonian([2.0], [hermitian_op])

    time = 7
    
    # run on IonQ simulator
    #tape = qml.tape.QuantumScript([qml.evolve(H, time)], [qml.probs(wires=[0, 1])])
    #result_ionq = dev.batch_execute([tape])

    # We multiply by -1 to add the expected negative sign
    pauli_decomp = qml.pauli_decompose(-1 * H.matrix())
    coeffs, ops = pauli_decomp.terms()
    H = qml.Hamiltonian(coeffs[::-1], ops[::-1])
    tape = qml.tape.QuantumScript([qml.TrotterProduct(H, time, n=1)], [qml.probs(wires=[0, 1])])
    simulator = qml.device("default.qubit", wires=2)
    result_simulator = qml.execute([tape], simulator)

    results_qiskit_statevector_sim = [
        0.10784311,
        0.45583129,
        0.09056136,
        0.34576424,
    ]

    assert np.allclose(
        result_simulator, results_qiskit_statevector_sim, atol=1e-2
    )
```

Major points to note:
* We're multiplying the matrix by `-1` before giving to `pauli_decompose` to account for the difference in convention of the time evolution
* We're reversing the order of the terms in the Hamiltonian before giving it as input to `qml.TrotterProduct`
* We're using `default.qubit` as the PL simulator. This one is just added because `default.qubit` and `lightning.qubit` should behave the same with `qml.TrotterProduct`, so it doesn't matter which simulator you use.

Hopefully this resolves all the remaining issues you have. I'm happy to support if you encounter any other issues :)
